fvm/transaction.go

```diff
@@ -70,65 +76,123 @@ func (i *TransactionInvocator) Process(
 	proc *TransactionProcedure,
 	st *state.State,
 ) error {
-	env, err := newEnvironment(ctx, vm, st)
-	if err != nil {
-		return err
-	}
-	env.setTransaction(proc.Transaction, proc.TxIndex)
 
-	location := common.TransactionLocation(proc.ID[:])
+	var err error
+	var env *hostEnv
+
+	numberOfRetries := 0
+	for numberOfRetries = 0; numberOfRetries < int(ctx.MaxNumOfTxRetries); numberOfRetries++ {
+		env, err = newEnvironment(ctx, vm, st)
+		// env construction error is fatal
+		if err != nil {
+			return err
+		}
+		env.setTransaction(proc.Transaction, proc.TxIndex)
+
+		location := common.TransactionLocation(proc.ID[:])
+
+		err = vm.Runtime.ExecuteTransaction(
+			runtime.Script{
+				Source:    proc.Transaction.Script,
+				Arguments: proc.Transaction.Arguments,
+			},
+			runtime.Context{
+				Interface: env,
+				Location:  location,
+			},
+		)
+
+		// break the loop
+		if !i.requiresRetry(err, proc) {
+			break
+		}
+
+		var blockHeight uint64
+		if ctx.BlockHeader != nil {
+			blockHeight = ctx.BlockHeader.Height
+		}
 
-	err = vm.Runtime.ExecuteTransaction(
-		runtime.Script{
-			Source:    proc.Transaction.Script,
-			Arguments: proc.Transaction.Arguments,
-		},
-		runtime.Context{
-			Interface: env,
-			Location:  location,
-		},
-	)
+		i.logger.Info().
+			Str("txHash", proc.ID.String()).
+			Uint64("blockHeight", blockHeight).
+			Int("retries_count", numberOfRetries).
+			Uint64("ledger_interaction_used", st.InteractionUsed()).
+			Msg("retrying transaction execution")
+
+		// reset error part of proc
+		// Warning right now the tx requires retry logic doesn't change
+		// anything on state but we might want to revert the state changes (or not commiting)
+		// if we decided to expand it furthur.
+		proc.Err = nil
+		proc.Logs = make([]string, 0)
+		proc.Events = make([]flow.Event, 0)
+		proc.ServiceEvents = make([]flow.Event, 0)
+		proc.Retried++
+	}
+
+	// (for future) panic if we tried several times and still failing
+	// if numberOfTries == maxNumberOfRetries {
+	// 	panic(err)
+	// }
 
 	if err != nil {
-		i.safetyErrorCheck(err)
+		var blockHeight uint64
```

Comment:
can't we just do this once, outside this if, on line 137, instead of twice, both here and 156?

Reply:
for sure, don't know why I did that in the first place

Reply:
fixed at e61f0ec